docs: improve website organization

tomMoral opened the pull request:
Related to doc improvements described in #1008 

## Summary of the changes

- remove GIF and replace by executable code snippet
- reorganized landing page with clickable boxes with links to motivation, tutorials, installation, and citation
- update appearance of API reference to be less cluttered
- remove tutorial dropdown in the navigation bar

## New `docs` workflow with `mike`

- [`mike`](https://github.com/jimporter/mike) allows us to have different `docs` versions in parallel
- we want to have one `dev` version showing the docs of the `main` branch (unreleased), and one `latest` version showing the docs of the latest release
- the website docs should always show the `latest` version

- whenever we push to `main`, the `dev` version is updated: 
    - `tutorials` and `examples` notebooks are converted into markdown and the docs files are created using `mkdocs build`
- whenever we publish a new release, the same thing happens, but the `latest` docs is updated (`-u` option in the command) to the new release and deployed on the website.

janfb commented:
Note: for building the docs site locally I needed to run the following commands in an empty condo env with `python=3.10`: 

```bash
pip install -e ".[doc,dev]"
pip install jupyter_contrib_nbextensions
pip install --upgrade notebook==6.4.12
pip uninstall traitlets 
pip install traitlets==5.9.0
pip uninstall ipython
pip install ipython==8.9.0
```
...
```bash
mkdocs serve
```

I believe this is due to our custom Jupyter notebook extensions (e.g., `jupyter_contrib_nbextensions`) bot being compatible with the new `notebook 7` release from last year. As a consequence, we need to downgrade `notebook`, `traitlets` and `ipython`.

janfb commented on docs/docs/index.md:
````diff
@@ -1,8 +1,21 @@
+# inference
+posterior = inference.build_posterior()
+posterior_samples = posterior.sample((1000,), x=x_o)
+```
````
I think it's time to remove the GIF. It just changes to fast and I believe it's not really helpful. I suggest to replace it with an actual code snippet.

tomMoral replied:
I like the idea of an actual code snipet! I think this helps onboarding a lot.
I suggest to add a dummy prior and simulator so this code is actually runnable, it helps a lot understanding the base concept of the toolkit.

tomMoral commented on docs/docs/contribute.md:
````diff
@@ -216,10 +216,3 @@ mkdocs serve
 and open a browser on the page proposed by `mkdocs`. Now, whenever you
 make changes to the markdown files of the documentation, you can see the results
 almost immediately in the browser.
-
-Note that the tutorials and examples are initially written in jupyter notebooks
-and then converted to markdown programatically. To do so locally, you should run
-```
-jupyter nbconvert --to markdown ../tutorials/*.ipynb --output-dir docs/tutorial/
-jupyter nbconvert --to markdown ../examples/*.ipynb --output-dir docs/examples/
-```
````
yes, with the `mkdocs-jupyter` extension.
However, this makes the rendering a bit slower with `mkdocs serve` as this is run each time you do an update on any file. I will see if it is possible to cache this.

tomMoral commented on docs/docs/index.md:
```diff
@@ -1,8 +1,21 @@
+
+# inference
+posterior = inference.build_posterior()
+posterior_samples = posterior.sample((1000,), x=x_o)
```
```suggestion
import torch
from sbi.inference import SNPE
from sbi.utils import BoxUniform

#define prior and simulator
prior = BoxUniform(torch.tensor([0]), torch.tensor([1]))
simulator = lambda theta: theta + 1

# simulation
theta = prior.sample((1000,))
x = simulator(theta)

# training
inference = SNPE(prior).append_simulations(theta, x)
inference.train()

# inference
x_o = torch.tensor([1.5])
posterior = inference.build_posterior()
posterior_samples = posterior.sample((1000,), x=x_o)
```

tomMoral commented on docs/docs/index.md:
````diff
@@ -86,48 +87,31 @@ The methods then proceed by
+or in a few lines for more flexibility:
 
-- **Contrastive Neural Ratio Estimation**<br>Benjamin Kurt Miller, Christoph Weniger, Patrick Forré (NeurIPS 2022) <br>[[PDF]](https://arxiv.org/pdf/2210.06170.pdf)
+```python
+inference = SNPE(prior=prior)
+_ = inference.append_simulations(theta, x).train()
+posterior = inference.build_posterior()
+```
````
I think this showcase the two different API so I would keep it.
But it is true it is more redundant.

tomMoral commented on docs/docs/reference/index.md:
```diff
@@ -0,0 +1,21 @@
+
+- [Inference](inference.md)
+  <br/>
+  *XXX*
```
I was not sure of what to put here, if you have suggestion that would be perfect.

tomMoral commented on docs/mkdocs.yml:
```diff
@@ -3,27 +3,26 @@ site_url: "https://sbi-dev.github.io/sbi/"
-        - Getting started: tutorial/00_getting_started_flexible.md
-        - Amortized inference: tutorial/01_gaussian_amortized.md
-        - Implemented algorithms: tutorial/16_implemented_methods.md
+        - Getting started: tutorials/00_getting_started_flexible.ipynb
+        - Amortized inference: tutorials/01_gaussian_amortized.ipynb
+        - Implemented algorithms: tutorials/16_implemented_methods.md
```
Yes I am not sure either which one is the right one, I will check this.

tomMoral reviewed (commented):
A few more comment, I will take another look at the `mkdocs-jupyter` extension in the afternoon

janfb commented:
@tomMoral I made some edits here: 

- shorter code snippet
- moved list publications / implemented algorithms back to `home` page
- remove getting started at the bottom of `home`
- use separate `index.md` for tutorials and API refs and remove the dropdown menu in the navigation bar

What do think of those?

And I have a couple of comments / questions: 
- at the moment it seems we have a copy of all notebooks in `docs/tutorials` and `docs/examples` and we convert them to markdown at build time
- I think this is not ideal: if we change the actual notebooks at the root level we need to update those as well, no?
- also, as you mentioned already, it takes about 10s at every build. 
- couldn't we convert them to markdown from the root level directly and do that with a separate command only when we deploy the website (like it was done before: https://github.com/sbi-dev/sbi/wiki/Release-workflow#update-the-documentation-to-the-new-version
- and the could wrap all the required commands into a GitHub action to have a release workflow?

tomMoral commented:
I finally got some time to take a look at this one.
I like the updates and I think the new look is more accessible.

I just pushed 2 small changes, one adding a  link to "get started" tutorial on the front page. Feel free to revert if you don't like it.

As for `mkdocs-jupyter`:
- The `docs/tutorials` and `docs/examples` are symlink to the root level ones so they are the same :)
- but indeed 10s is slow for serving.

This is really as you prefer for the workflow.
I would tend to say that having a fast iteration path for updating the doc is better so I will try and see if we can automatize calling `nbconvert`.

janfb commented:
Hi @tomMoral 

thanks for the update! It looks good. 

I made a couple of smaller changes:

- edited the landing page: updated code snippet, updated implemented methods
- merge `main` into this branch
- update tutorials of implemented methods
- update credits

Two points remain: 

1. It seems when clicking on the tutorial notebooks and looking at them on the website in markdown, they do not show the code but only the markdown content and the output. See for example "Getting started". So there seems to be something off here. 
2. Workflow: I think it would be good if we have GH workflow that is launched whenever me make a new release, i.e., not on every push to `main`. The reason is that we want to keep the docs in sync with the pypi version of `sbi`, not with `main`. I would also prefer to separate the notebook conversion from the website building. In other words, we include a manual call to `nbconvert` into the workflow that deploys new docs upon a new release, but when I work on the website locally it not re-convert the notebooks with every change to the website. Does that make sense?

janfb commented:
@tomMoral In the last commit I made a couple of naive changes to the workflow file, all up for discussion: 

- action is issued upon a published release
- converts notebooks from the source dir into the `docs/` into markdown, just like we did manually before
- directly deploy the docs onto the website.

Does that make sense?

tomMoral commented:
- I removed `mkdocs-jupyter`, which we are not using anymore.
- For the doc versioning, it seems that [`mike`](https://github.com/jimporter/mike) is a package to have multiple versions (one dev and one per release, with a `latest` alias which would be the default).
- I updated the workflow to use it for now.

janfb commented:
TODO: 

- [ ] make local backup of `docs` folder
- [ ] delete all previous docs using `mike delete --all
- [ ] checkout on tag of the last release `v0.22.0`
- [ ] build the docs using `mike deploy 0.22.0 latest -u --push`
- [ ] set default docs using `mike set-default latest`

janfb commented:
@michaeldeistler can you please give this a final review? To do so

- checkout locally and install the new `docs` dependencies: `pip install -e ".[doc]"`
- run `mike serve`
- see website on localhost.

See also PR description on top which summarizes all the changes.
